File: packages/auditcore_bpmn/src/auditcore_bpmn/collection/excerpt.py
"""Auszug eines Diagramms für die Sammlung (Prozesse, Verweise, Kennzahlen, Schlüsselindex)."""

from __future__ import annotations

from ..extensions import DiagramInfo, Extensions
from ..model import ACTIVITIES, BpmnDocument
from .model import KEY_KINDS, DiagramExcerpt

_REFERENCE_KINDS = ("prueffeld", "feststellung_ref", "register")


class _KeyIndex:
    def __init__(self) -> None:
        self.keys: dict[str, dict[str, list[str]]] = {kind: {} for kind in KEY_KINDS}

    def add(self, kind: str, value: str | None, element_id: str) -> None:
        """Nimmt einen Schlüsselwert eines Elements auf."""
        if value:
            ids = self.keys[kind].setdefault(str(value).strip(), [])
            if element_id not in ids:
                ids.append(element_id)

    def add_extensions(self, element_id: str, ext: Extensions) -> None:
        """Schlüssel aus den Erweiterungen eines Elements."""
        for reference in ext.audit_references:
            self.add("ka", reference.key_requirement, element_id)
            self.add("bk", reference.assessment_criterion, element_id)
        for finding in ext.findings:
            self.add("ka", finding.key_requirement, element_id)
            self.add("bk", finding.assessment_criterion, element_id)
            self.add("feststellung_ref", finding.reference, element_id)
        for link in ext.cross_references:
            if link.kind in _REFERENCE_KINDS:
                self.add(link.kind, link.key, element_id)
        if ext.actor is not None:
            self.add("rolle", ext.actor.role, element_id)
        for marker in ext.markers:
            self.add("kennzeichen", marker.type, element_id)

    def add_info(self, owner: str, info: DiagramInfo) -> None:
        """Schlüssel aus den Diagramm-Infos."""
        for reference in info.audit_references:
            self.add("ka", reference.key_requirement, owner)
            self.add("bk", reference.assessment_criterion, owner)
        for link in info.cross_references:
            if link.kind in _REFERENCE_KINDS:
                self.add(link.kind, link.key, owner)

    def result(self) -> dict[str, dict[str, list[str]]]:
        """Index ohne leere Arten."""
        return {kind: values for kind, values in self.keys.items() if values}


def excerpt_from_document(document: BpmnDocument) -> DiagramExcerpt:
    """Auszug für die Sammlung aus einem gelesenen Diagramm."""
    index = _KeyIndex()
    for element in document:
        index.add_extensions(element.id, element.extensions)
    if document.diagram_info is not None:
        index.add_info(document.diagram_info_owner or "diagramm", document.diagram_info)
    activities = [e for e in document.flow_nodes if e.type in ACTIVITIES]
    return DiagramExcerpt(
        process_ids=[p.id for p in document.processes],
        calls=[(e.id, e.called_element) for e in document.by_type("callActivity") if e.called_element],
        link_throws=[(e.id, e.link_name) for e in document.by_type("intermediateThrowEvent") if e.link_name],
        link_catches=[(e.id, e.link_name) for e in document.by_type("intermediateCatchEvent") if e.link_name],
        activities=len(activities),
        activities_with_legal_basis=sum(1 for e in activities if e.extensions.legal_bases),
        keys=index.result(),
    )

File: packages-js/bpmn-flowaudit/src/collection/excerpt.ts
/**
 * Excerpt of a diagram for the collection: processes, calls, link events,
 * task counts and the index of stable domain keys.
 */

import { isActivity, localType, type ProcessModel } from '../model/processModel'
import type { AuditReference, CrossReference } from '../schema/types'
import { emptyExcerpt, type DiagramExcerpt, type KeyKind } from './collectionData'

class KeyIndex {
  readonly keys: Partial<Record<KeyKind, Record<string, string[]>>> = {}

  add(kind: KeyKind, value: string | undefined, elementId: string): void {
    const key = value?.trim()
    if (!key) return
    const byValue = (this.keys[kind] ??= {})
    const ids = (byValue[key] ??= [])
    if (!ids.includes(elementId)) ids.push(elementId)
  }

  references(list: AuditReference[], elementId: string): void {
    for (const ref of list) {
      this.add('ka', ref.keyRequirement, elementId)
      this.add('bk', ref.assessmentCriterion, elementId)
    }
  }

  crossReferences(list: CrossReference[], elementId: string): void {
    for (const ref of list) {
      if (ref.kind === 'prueffeld' || ref.kind === 'feststellung_ref' || ref.kind === 'register') this.add(ref.kind, ref.key, elementId)
    }
  }
}

function indexKeys(model: ProcessModel): KeyIndex {
  const index = new KeyIndex()
  for (const element of model.elements) {
    const ext = element.extensions
    index.references(ext.auditReferences, element.id)
    for (const finding of ext.findings) {
      index.references([{ keyRequirement: finding.keyRequirement, assessmentCriterion: finding.assessmentCriterion }], element.id)
      index.add('feststellung_ref', finding.reference, element.id)
    }
    index.crossReferences(ext.crossReferences, element.id)
    index.add('rolle', ext.actor?.role, element.id)
    for (const marker of ext.markers) index.add('kennzeichen', marker.type, element.id)
  }
  if (model.info) {
    const where = model.mainId ?? 'diagramm'
    index.references(model.info.auditReferences ?? [], where)
    index.crossReferences(model.info.crossReferences ?? [], where)
  }
  return index
}

function linkEvents(model: ProcessModel, type: string): [string, string][] {
  return model.elements
    .filter((el) => localType(el.type) === type && el.linkName)
    .map((el): [string, string] => [el.id, el.linkName as string])
}

export function excerptFromModel(model: ProcessModel): DiagramExcerpt {
  const activityList = model.elements.filter((el) => isActivity(el.type))
  return {
    ...emptyExcerpt(),
    processIds: model.elements.filter((el) => el.type === 'bpmn:Process').map((el) => el.id),
    calls: model.elements.filter((el) => el.calledElement).map((el): [string, string] => [el.id, el.calledElement as string]),
    linkThrows: linkEvents(model, 'intermediateThrowEvent'),
    linkCatches: linkEvents(model, 'intermediateCatchEvent'),
    tasks: activityList.length,
    tasksWithLegalBasis: activityList.filter((el) => el.extensions.legalBases.length > 0).length,
    keys: indexKeys(model).keys,
  }
}

/** Element ids of the diagram model matching a key (for highlighting). */
export function elementsForKey(model: ProcessModel, kind: KeyKind, value: string): string[] {
  return indexKeys(model).keys[kind]?.[value.trim()] ?? []
}
